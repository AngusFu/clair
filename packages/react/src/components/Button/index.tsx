import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const types = [
  "primary",
  "default",
  "warning",
  "danger",
  "success",
  "ghost"
] as const;
export type Type = typeof types[number];
const PType = PropTypes.oneOf([...types]);

const sizes = ["large", "normal", "small"] as const;
export type Size = typeof sizes[number];
const PSize = PropTypes.oneOf([...sizes]);

const htmlTypes = ["button", "reset", "submit"] as const;
export type HtmlType = typeof htmlTypes[number];
const PHtmlType = PropTypes.oneOf([...htmlTypes]);

// handle "detail" property of custom event
type ButtonEvent = React.MouseEvent<HTMLButtonElement> &
  Partial<Pick<CustomEvent, "detail">>;
type EventHandler = React.EventHandler<ButtonEvent>;

export interface ButtonProps {
  size?: Size;
  type?: Type;
  htmlType?: HtmlType;
  disabled?: boolean;
  loading?: boolean;
  block?: boolean;
  className?: string;
  style?: React.CSSProperties;
  children?: React.ReactNode;
  onClick?: EventHandler;
  onSubmit?: EventHandler;
  onReset?: EventHandler;
  forwardRef?: React.RefObject<HTMLButtonElement>;
}

export const Button: React.FC<ButtonProps> = props => {
  const {
    size = "normal",
    type = "default",
    htmlType = "button",
    disabled = false,
    loading = false,
    block = false,
    children,
    forwardRef,
    className,
    ...rest
  } = props;

  const className2Use: string = classNames("c-button", className, {
    [`c-button--${type}`]: types.some(t => t === type),
    [`c-button--${size}`]: sizes.some(s => s === size),
    "c-button--block": block,
    "c-button--loading": loading
  });
  return (
    <button
      ref={forwardRef}
      className={className2Use}
      disabled={disabled}
      type={htmlType}
      {...rest}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  size: PSize,
  type: PType,
  htmlType: PHtmlType,
  disabled: PropTypes.bool,
  block: PropTypes.bool,
  loading: PropTypes.bool,
  className: PropTypes.string,
  style: PropTypes.object,
  onClick: PropTypes.func,
  onSubmit: PropTypes.func,
  onReset: PropTypes.func,
  children: PropTypes.node,
  forwardRef: PropTypes.shape({
    current: PropTypes.any
  })
} as PropTypes.ValidationMap<ButtonProps>;

Button.displayName = "Button";
