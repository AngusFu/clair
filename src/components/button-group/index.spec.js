import { mount } from '@vue/test-utils'
import Button from '../button/index.vue'
import ButtonGroup from './index.vue'

it('should works with size', () => {
  const wrapper = mount(ButtonGroup, {
    propsData: { size: 'md' },
    stubs: {
      'c-button': Button
    },
    slots: {
      default: [
        `
<c-button primary outline>A</c-button>
<c-button primary outline>B</c-button>
<c-button primary outline>C</c-button>
`
      ]
    }
  })

  expect(wrapper.classes()).toContain('is-md')
  expect(wrapper.findAll(Button).length).toBe(3)
  expect(wrapper).toMatchSnapshot()
})
