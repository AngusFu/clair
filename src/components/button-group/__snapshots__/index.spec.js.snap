// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should works with size 1`] = `
<div class="c-button-group is-md"><button type="button" class="c-button c-button--primary c-button--outline">
    <!----><span>A</span></button> <button type="button" class="c-button c-button--primary c-button--outline">
    <!----><span>B</span></button> <button type="button" class="c-button c-button--primary c-button--outline">
    <!----><span>C</span></button></div>
`;
